---
title: "RATs: Output and Plots"
author: "Kimon Froussios"
date: "16/02/2017"
output: 
  html_document: 
    fig_caption: no
    keep_md: yes
    theme: readable
    toc: yes
vignette: >
  %\VignetteIndexEntry{RATs Output and Plots}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

***

Set up an example.

```{r}
library(rats)

# Simulate some data.
simdat <- sim_sleuth_data(cnames = c("controls", "patients")) 
# For convenience let's assign the contents of the list to separate variables.
myslo <- simdat$slo
myannot <- simdat$annot

# Call DTU
mydtu <- call_DTU(annot = myannot, slo = myslo, name_A = "controls", name_B = "patients", 
                  varname= "condition", verbose= FALSE,
                  description="Comparison of two conditions using a simulated sleuth object 
                    for the purposes of the tutorial. Simulated using built-in functionality 
                    of RATs.")
```


***


# Quick results

For your convenience, RATs provides a couple of functions to give you a quick summary of your results.
However, we do recommend you become familiar with the actual results structure and content, so that you
can judge the quality of the DTU calls and trace the reasons behind the classification of each gene or transcript.

The `dtu_summary()` function lists the total number of genes and transcripts for each of 3 categories:

```{r}
# A really simple tally of the outcome.
print( dtu_summary(mydtu) )
```

* DTU:  There is significant change in isoform ratios.
* non-DTU:  No significant change.
* NA:  Not applicable. Genes/transcripts with abundance below the noise threshold, or where the gene has only one transcript.

The `get_dtu_ids()` function lists the actual identifiers per category, instead of the numbers in each category.
The categories are the same as those in `dtu_summary()` above.

```{r}
# Gene and transcript IDs corresponding to the tally above.
ids <- get_dtu_ids(mydtu)

# Contents
print( names(ids) )

# DTU positive genes.
print( ids[["dtu-genes"]] )
```

The ID lists obtained with `get_dtu_ids()` are *ordered by effect size* (`Dprop`).


***


# Output structure

The output of RATs is a list containing 4 elements:

```{r}
print( names(mydtu) )
```


## Parameters

`Parameters` is a list that contains information about the data and the settings for a particular run.

```{r}
# Parameter list's elements.
print( names(mydtu$Parameters) )
```

1. `description` - (str) Free-text description of the run. It is useful to record data sources, annotation source and version, experimental parameters...
2. `time` - (str) Date and time for the run. This does not represent the exact time the run started.
3. `rats_version` - (str) The version of RATs.
4. `R_version` - (str) The version of R (including OS architecture).
5. `var_name` - (str) The value passed to the `varname` parameter.
6. `cond_A` & `cond_B` - (str) The values passed to the `name_A` and `name_B` parameters.
7. `data_type` - (str) The format of the input data.
8. `num_replic_A` & `num_replic_B` - (int) The number of samples in each condition.
9. `num_genes` - (int) The number of genes in the provided annotation.
10. `num_transc` - (int) The number of transcripts in the provided annotation.
11. `tests` - (str) The value passed to the `testmode` parameter.
12. `p_thresh` - (num) The value passed to the `p_thresh` parameter.
13. `abund_thresh` - (num) The value passed to the `abund_thresh` parameter.
14. `dprop_thresh` - (num) The value passed to the `dprop_thresh` parameter.
15. `quant_reprod_thresh` - (num) The value passed to the `qrep_thresh` parameter.
16. `quant_boot` - (bool) The value passed to the `qboots` parameter.
17. `quant_bootnum` - (int) The value passed to the `qbootnum` parameter.
18. `rep_reprod_thresh` - (num) The value passed to the `rrep_thresh` parameter.
19. `rep_boots` - (bool) The value passed to the `rboot` parameter.
20. `rep_bootnum` - (int) The number of replicate bootstrapping iterations (usually  M*N, where M and N are the number of samples in the two conditions).
21. `rep_reprod_as_crit` - (bool) The value passed to the `rrep_as_crit` parameter.

**Note:** If bootstraps are disabled, the bootstrap-related fields may have `NA` values despite any values that were passed to their respective parameters.


## Genes

`Genes` is a [data.table](https://cran.r-project.org/web/packages/data.table/) with many fields, listing 
results at the gene level.

The G-test is used for gene-level calls. The G-test is designed to compare a set of counts against a set of ratios. To use it with two sets of
counts, we do the comparison both ways, using each condition in turn for each role. Thus, the table contains two sets of results,
marked with the `AB` and `BA` suffixes. It is possible to get a DTU-positive result here, even when all the individual transcripts
are predicted as DTU-negative, as the G test takes into account more information than the individual transcript proportion equality tests.

```{r}
# Genes table's fields.
print( names(mydtu$Genes) )
```

The first few columns show the result of each decision step. The remaining columns list the values based on which the decisions were made.
Pseudo-code formulas are shown to help understand how the different fields interact when making decisions.

1. `parent_id` - (str) Gene identifier.
2. `elig` - (bool) Eligible for testing. Whether the gene met the pre-filtering criteria. `= (elig_transc >= 2)`
3. `sig` - (bool) Statistically significant. `= (pvalAB_corr < Parameters$p_thresh) & (pvalBA_corr < Parameters$p_thresh)`
4. `elig_fx` - (bool) Eligible effect size. Whether at least one of the isoforms meets the effect size criterion. `= any(Transcript[parent_id, elig_fx])`
5. `quant_reprod` - (bool) Quantification reproducible.
For positive DTU, `= (quant_dtu_freq >= Parameters$quant_reprod_thresh)`, for non-DTU `= (quant_dtu_freq <= 1 - Parameters$quant_reprod_thresh)`
6. `rep_reprod` - (bool) Replication reproducible.
For positive DTU, `= (rep_dtu_freq >= Parameters$rep_reprod_thresh)`, for non-DTU `= (rep_dtu_freq <= 1 - Parameters$rep_reprod_thresh)`
7. `DTU` - (bool) Differential Transcript Usage. `= (sig & elig_fx & quant_reprod)`, or if `rrep_as_crit==TRUE`, then `= (sig & elig_fx & quant_reprod & rep_reprod)`
8. `transc_DTU` - (bool) Aggregated from `Transcripts$DTU`: At least one isoform was individually reported as DTU. `= any(Transcript[parent_id, DTU])`
9. `known_transc` - (int) Number of known trascripts for this gene according to the given annotation.
10. `detect_trancs` - (int) Number of detected (expressed) transcripts in the given dataset.
11. `elig_transc` - (int) Number of eligible transcripts, aggregated from `Transcripts$elig`.
12. `pvalAB` & `pvalBA` - (num) G test p-value using in turn each condition as the reference ratios.
13. `pvalAB_corr` & `pvalBA_corr` - (num) Multiple testing corrected p-value. `pvalAB_corr= p.adjust(pvalAB, Parameters$correction)`
14. `quant_p_meanAB` & `quant_p_meanBA` - (num) Mean p-value across quantification bootstraps.
15. `quant_p_stdevAB` & `quant_p_stdevBA` - (num) Standard deviation of p-values across quantification bootstraps.
16. `quant_p_minAB` & `quant_p_minBA` - (num) Minimum observed p-value across quantification bootstraps.
17. `quant_p_maxAB` & `quant_p_maxBA` - (num) Maximum observed p-value across quantification bootstraps.
18. `quant_na_freq` - (num) Fraction of quantification iterations in which `DTU` could not be determined (not meeting pre-filtering criteria).
19. `quant_dtu_freq` - (num) Fraction of replicate iterations that support a positive `DTU` classification.
20. `rep_p_meanAB` & `rep_p_meanBA` - (num) Mean p-value across replication bootstraps.
21. `rep_p_stdevAB` & `rep_p_stdevBA` - (num) Standard deviation of p-values across replication bootstraps.
22. `rep_p_minAB` & `rep_p_minBA` - (num) Minimum observed p-value across replication bootstraps.
23. `rep_p_maxAB` & `rep_p_maxBA` - (num) Maximum observed p-value across replication bootstraps.
24. `rep_na_freq` - (num) Fraction of replication iterations in which `DTU` could not be determined (not meeting pre-filtering criteria).
25. `rep_dtu_freq` - (num) Fraction of replication iterations that support a positive `DTU` classification.

**Note:** The fields reporting on the bootstraps will not be shown when bootstrapping is disabled.


## Transcripts

`Transcripts` is a [data.table](https://cran.r-project.org/web/packages/data.table/) with many fields, listing 
results at the transcript level. For your convenience, the gene-level DTU calls are also included here.

The proportions test is used for the transcript-level calls. Changes can be attributed to specific transcripts,
but, as it uses less information, the test has a higher detection threshold and requires larger changes than the gene-level test.

```{r}
# Transcripts table's fields.
print( names(mydtu$Transcripts) )
```

1. `target_id` - (str) Transcript identifier.
2. `parent_id` - (str) Gene identifier.
3. `elig_xp` - (bool) Eligible expression level. `= (meanA >= Parameters$abund_thresh | meanB >= Parameters$abund_thresh)`
4. `elig` - (bool) Eligible for testing (meets the noise threshold and at least one other isoform is expressed). `= (elig_xp & totalA != 0 & totalB != 0 & (sumA != totalA | sumB != totalB))`
5. `sig` - (bool) Statistically significant. `= (pval_corr < Parameters$p_thresh)`
6. `elig_fx` - (bool) Eligible effect size. Proxy for biological significance. `= (Dprop > Parameters$dprop_thresh)`
7. `quant_reprod` - (bool) Quantification reproducible.
For positive DTU, `= (quant_dtu_freq >= Parameters$quant_reprod_thresh)`, for non-DTU `= (quant_dtu_freq <= 1 - Parameters$quant_reprod_thresh)`.
8. `rep_reprod` - (bool) Replication reproducible.
For positive DTU, `= (rep_dtu_freq >= Parameters$rep_reprod_thresh)`, for non-DTU `= (rep_dtu_freq <= 1 - Parameters$rep_reprod_thresh)`.
9. `DTU` - (bool) The transcript's proportion changed significantly. Calculated as: `DTU = (sig & elig_fx & conf)`.
10. `gene_DTU` - (bool) Expanded from `Genes$DTU`. Indicates that the gene as a whole shows significant change in isoform ratios. `= (Genes[parent_id, DTU])`
11. `meanA` and `meanB` - (num) The mean abundance across replicates, for each condition.
12. `stdevA` and `stdevB` - (num) The standard deviation of the abundance across the replicates.
13. `sumA` and `sumB` - (num) The sum of the abundances across replicates. This is the value used for the tests, so that replication level informs the significance.
14. `totalA` and `totalB` - (num) The total abundance for the gene: `totalA= sum(transcripts[parent_id, sumA])`
15. `propA` and `propB` - (num) The proportion of the gene expression owed to this transcript. `propA= sumA/totalA`
16. `Dprop` - (num) The difference in the proportion of the transcript between the two conditions (effect size). `= (probB - propA)`
17. `pval` - (num) The proportion equality test P-value for the transcript.
18. `pval_corr` - (num) Multiple testing corrected p-value. `= p.adjust(pval, Parameters$correction)`
19. `quant_p_mean` - (num) Mean p-value across quantification bootstraps.
20. `quant_p_stdev` - (num) Standard deviation of p-values across quantification bootstraps.
21. `quant_p_min` - (num) Minimum observed p-value across quantification bootstraps.
22. `quant_p_max` - (num) Maximum observed p-value across quantification bootstraps.
23. `quant_na_freq` - (num) Fraction of quantification iterations in which `DTU` could not be determined (not meeting pre-filtering criteria).
24. `quant_dtu_freq` - (num) Fraction of quantification iterations that support a positive `DTU` classification.
25. `rep_p_mean` - (num) Mean p-value across replication bootstraps.
26. `rep_p_stdev` - (num) Standard deviation of p-values across replication bootstraps.
27. `rep_p_min` - (num) Minimum observed p-value across replication bootstraps.
28. `rep_p_max` - (num) Maximum observed p-value across replication bootstraps.
29. `rep_na_freq` - (num) Fraction of replication iterations in which `DTU` could not be determined (not meeting pre-filtering criteria).
30. `rep_dtu_freq` - (num) Fraction of replication iterations that support a positive `DTU` classification.

**Note** The fields reporting on the bootstraps will not be shown when bootstrapping is disabled.


## Abundances
 
`Abundances` is a list of two `data.table`s, one for each condition. Each transcript is represented by a single abundance value per replicate, 
so if bootstrapped data was used, these values are the means across iterations. If plain abundances were provided as input, then `Abundances` 
essentially contains the input data. These abundances are included in the output because they are required for some of RATs' plotting options.


```{r}
# Elements of ReplicateData
print( names(mydtu$Abundances) )
```

1. `condA` - (num) The transcript abundances in the first condition.
2. `condB` - (num) The transcript abundances in the second condition.


***


# Visualisation of results


The output object's tables provide a host of information. The `rats` package also includes some visualisation aides.


## Isoform abundance changes for a given gene

This function allows you to visualise what's going on in any particular gene. Both the absolute counts and the relative 
proportions are plotted for each transcript. This is a very useful function for inspecting a gne if interest. It enables 
quick visual evaluation of the dispersion of the replicate measurements, the magnitude of the proportion change, the 
presence of outliers, and the consistency among the replicates.

```{r}
# Proportion and count changes for all the transcripts of the "MIX6" gene.
plot_gene(mydtu, "MIX6", style="lines")  # default
```

Several styles of this plot are available. The recommended style is set as default. The other options represent older 
styles that have been kept for backwards compatibility, but may get deprecated eventually: 

```{r eval=FALSE}
plot_gene(mydtu, "MIX6", style="points")
plot_gene(mydtu, "MIX6", style="rainbow")
plot_gene(mydtu, "MIX6", style="merged")
plot_gene(mydtu, "MIX6", style="dashed")
```


## Plots of overall run

Our simulated dataset is too small to properly demonstrate what these plots typically look like.
So each one is accompanied by a static image of the same plot created with a real and much larger dataset.

Several of these plots are **likely to display warnings** about missing or non-finite values. These are due to the 
presence of `NA` in the tables, where entries did not pass the thresholds, and can be ignored.

Possibly the most common plot in differential expression is the volcano plot, which plots the effect size against 
the statistical significance. As it is difficult to define a single p-value and a single effect size at the gene level,
the volcano can only be plotted at the transcript level.

```{r}
# Proportion change VS significance.
plot_overview(mydtu, type="volcano")
```

This is what it looks like on a larger dataset:
![Dprop VS sig](./fig/volcano.jpg)

The next command plots the largest change in proportion seen within each gene, against the number of genes showing 
such change. This is a way to inspect what effect sizes are present in the data. As an additional layer of information,
they are colour-coded by their DTU call.

```{r}
# Distribution of maximum proportion change.
plot_overview(mydtu, type="maxdprop")
```

This is what it looks like on a larger dataset:
![Max Dprop](./fig/maxdprop.jpg)


### Interactive plots

If you prefer picking points from a plot than sorting through tables, the volcano plot is also available through 
a `shiny` app, that brings up the relevant abundance changes plot for any point in the volcano plot.

1. By hovering over points on the volcano plot in the app, you can see the respective transcript identifier(s). 
2. Clicking will pull up information on the effect size, significance and confidence of the point(s), as well as 
the respective isoform abundance changes plot for the point nearest to the click.

```{r, eval=FALSE}
# Start the interactive volcano plot.
plot_shiny_volcano(mydtu)
```

This is what it looks like for the example data (remember that the emulated data example has very few transcripts).

![Transc Conf VS DTU](./fig/shiny_screenshot.png)

You will need to close down the app to return to your R terminal.


## Plot customisation

You can save any of the plots as a `ggplot2` object and use [ggplot2](http://ggplot2.org) manipulations on it, such as changing the axis scales.
Other `ggplot2` customisations include the axis tick marks, axis values, labels, titles, colours... Consult the [ggplot2](http://ggplot2.org)
documentation for more help on these.

```{r}
library(ggplot2)

myplot <- plot_overview(mydtu, "volcano")
myplot  # display

# Change title. 
myplot2 <- myplot + ggtitle("MY EPIC TITLE")
myplot2
```


***


# Contact information

The `rats` R package was developed within [The Barton Group](http://www.compbio.dundee.ac.uk) at [The University of Dundee](http://www.dundee.ac.uk)
by Dr. Kimon Froussios, Dr. Kira Mourão and Dr. Nick Schurch.

To **report problems** or **ask for assistance**, please raise a new issue [on the project's support forum](https://github.com/bartongroup/Rats/issues).
Providing a *reproducible working example* that demonstrates your issue is strongly encouraged to help us understand the problem. Also, be sure 
to **read the vignette(s)**, and browse/search the support forum before posting a new issue, in case your question is already answered there.

Enjoy!

![](./fig/rats.png)


