---
title: "RATs FDR"
author: "Kimon Froussios"
date: "21/02/2018"
output:
  html_document:
    keep_md: no
    theme: readable
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Premise

Measuring FDR of RATs for different replication levels (3-8 per condition), using 100 permutations of the combination of input samples per replication level.
The data are 17 replicates of Arabidopsis Colombia-0. Library sizes are 70-140M paired-end 100bp reads, prepared with the Illumina ribo-minus protocol. Quantification was done with Salmon 0.7.1 and the TPM values were scaled up to an appropriate average library size of 100M. The RATs version used was 0.6.0.

RATs output is summarized in a 5-dimensional R array per iteration, with dimensions [test type, reproducibility threshold, effect size threshold, 
significance cutoff, abundance threshold] and contained in a list.
The ranges of values used are:

```{r}
replevels <- as.character(c(3,4,5,6,7,8))
abund_range <- as.character(c(0, 1, 5, 10, 50, 100))
p_range <- as.character(c(0, 0.001, 0.005, 0.01, 0.05, 0.1))
dprop_range <- as.character(c(0, 0.01, seq(0.02, 0.3, 0.02), seq(0.4, 1, 0.2)))
qrep_range <- as.character(c(seq(0, 0.8, 0.2), seq(0.9, 1, 0.01)))
tests <- c("gene", "transc")

# Default thresholds:
dp <- "0.05"  # Significance
dd <- "0.2"   # Difference in proportions
dq <- "0.95"  # Quantification reproducibility
da <- "5"    # Abundance

# Samples.
reps <- c(1,2,3,4,5,6,7,8,9,10,12,13,14,15,16,17)  # 11 was porly corelated with the others due to inefficient rRNA removal.
```


# Set up

```{r}
library(parallel)
library(ggplot2)

# Theme for plots.
gth <- theme(axis.line.x= element_line(),
	         axis.line.y= element_line(),
	         strip.background= element_rect(fill= "grey95"),
	         strip.text.y= element_text(size= rel(1.2)),
	         strip.text.x= element_text(size= rel(1.1)),
	         panel.grid.major= element_blank(),
	         panel.grid.minor= element_blank(),
	         panel.background= element_rect(fill = "white"),
	         legend.key = element_rect(fill = 'white')
)

# Paths.
# base_dir <- "."
base_dir <- "/Volumes/kfroussios/PROJECTS/rats"
results_dir <- "FDR"

# Files
# bootfiles <- lapply(replevels,function(rl) {
# 	dir(file.path(base_dir, results_dir, rl))
# })
bootfiles <- lapply(replevels,function(rl) {
	sapply(seq.int(1,100,1), function(b) {
		file.path(base_dir, results_dir, rl, paste0(b, ".fp.RDS"))
	})
})
names(bootfiles) <- replevels
bootnum <- length(bootfiles[[1]])

# Number of genes and transcripts in the annotation.
load(file.path(base_dir, "genome/Araport11_t2g.rda"))
print( c(length(unique(araport11$parent_id)), length(araport11$target_id)) )
```


## Load data

```{r}
# A list of rep levels, containing 100 items per level, each item being a 5D array.
mega <- lapply(replevels, function (rl) { 
			return( mclapply(bootfiles[[rl]], function(bf) {
				bd <- readRDS(bf)
				return(bd[["dtu"]])
			}, mc.cores=7) )
		})
names(mega) <- replevels

elig <- lapply(replevels, function (rl) { 
			return( mclapply(bootfiles[[rl]], function(bf) {
				bd <- readRDS(bf)
				e <- bd[["elig"]]
				row.names(e) <- as.character(e$abund)
				return(e)
			}, mc.cores=7) )
		})
names(elig) <- replevels
```

# Explore


## Number of tested genes/transcripts

```{r}
# Collect number of eligible genes into ggplot-able dataframe.
eligens <- lapply(seq.int(1,bootnum,1), function(b) {
	e <- as.data.frame(t(as.data.frame( lapply(replevels, function(rl) {
		return( elig[[rl]][[b]][["eligen"]]  )
	}) )))
	e <- data.frame("elig"= unlist(e[seq.int(1, length(abund_range),1)]),
					"abund"= rep(as.numeric(abund_range), rep(length(replevels), length(abund_range))),
					"rep"= rep(replevels, length(abund_range)),
					"boot"= c(b),
					"type"= "genes")
	return(e)
})
eligens <- do.call("rbind", eligens)

# Collect number of eligible transcripts into ggplot-able dataframe.
elitras <- lapply(seq.int(1,bootnum,1), function(b) {
	e <- as.data.frame(t(as.data.frame( lapply(replevels, function(rl) {
		return( elig[[rl]][[b]][["elitra"]]  )
	}) )))
	e <- data.frame("elig"= unlist(e[seq.int(1, length(abund_range),1)]),
					"abund"= rep(as.numeric(abund_range), rep(length(replevels), length(abund_range))),
					"rep"= rep(replevels, length(abund_range)),
					"boot"= c(b),
					"type"= "transcripts")
	return(e)
})
elitras <- do.call("rbind", elitras)

# eli <- rbind(eligens, elitras)

print( ggplot(eligens, aes(y=elig, x=rep)) +
	facet_grid(. ~ abund, scales="free") +
	geom_boxplot(outlier.alpha=0, fill="blue", alpha=0.1) +
	#geom_jitter(width=0.2, alpha=0.3, size=rel(0.8)) +
	guides(fill="none") +
	ggtitle("Eligible genes by replication level for different abundance thresholds") +
	xlab("Replicates per condition") +
	ylab("Number of eligible genes") +
	scale_y_continuous(expand=c(0,0)) +
	coord_cartesian(ylim=c(0,11000)) +
	theme(axis.line.x=element_line())
)

print( ggplot(elitras, aes(y=elig, x=rep)) +
	facet_grid(. ~ abund, scales="free") +
	geom_boxplot(outlier.alpha=0, fill="blue", alpha=0.1) +
	#geom_jitter(width=0.2, alpha=0.3, size=rel(0.8)) +
	guides(fill="none") +
	ggtitle("Eligible transcripts by replication level for different abundance thresholds") +
	xlab("Replicates per condition") +
	ylab("Number of eligible transcripts") +
	scale_y_continuous(expand=c(0,0)) +
	coord_cartesian(ylim=c(0,33000)) +
	theme(axis.line.x=element_line())
)

rm(eligens, elitras)
```

The number of eligible features sets the background against which FDR is measured. It is quite stable across replication levels for a wide range of values.
When there is no abundance threshold (threshold is 0), the number of eligible features initially increases with the number of transcripts, as additional replicates increase the chance of at least one read being assigned to a feature.

## FDR: default parameters.

```{r}
# Gene-level results.
FDRsg <- as.data.frame( lapply(replevels, function(rl) {
	FDR <- unlist( lapply(seq.int(1,bootnum,1), function(b) {
		return( mega[[rl]][[b]]["gene", dq, dd, dp, da] / elig[[rl]][[b]][da, "eligen"] )
	}) )
	return(FDR)
}) )
names(FDRsg) <- replevels
# Reshape into single FP variable with sample labels, to facilitate plotting.
FDRsg <- data.frame("FDR"= c(FDRsg[["3"]], FDRsg[["4"]], FDRsg[["5"]], FDRsg[["6"]], FDRsg[["7"]], FDRsg[["8"]]),
				   "Reps"= as.character(rep(replevels, c(bootnum, bootnum, bootnum, bootnum, bootnum, bootnum))) )
FDRsg["testtype"] = c("genes")

# Transcript-level results.
FDRst <- as.data.frame( lapply(replevels, function(rl) {
	FDR <- unlist( lapply(seq.int(1,bootnum,1), function(b) {
		return( mega[[rl]][[b]]["transc", dq, dd, dp, da] / elig[[rl]][[b]][da, "elitra"] )
	}) )
	return(FDR)
}) )
names(FDRst) <- replevels
# Reshape into single FP variable with sample labels, to facilitate plotting.
FDRst <- data.frame("FDR"= c(FDRst[["3"]], FDRst[["4"]], FDRst[["5"]], FDRst[["6"]], FDRst[["7"]], FDRst[["8"]]),
				   "Reps"= as.character(rep(replevels, c(bootnum, bootnum, bootnum, bootnum, bootnum, bootnum))) )
FDRst["testtype"] = c("transcripts")

# Plot.
fp_default <- rbind(FDRsg, FDRst)

print( 
	ggplot(fp_default, aes(x=Reps, y=FDR)) +
		facet_grid(. ~ testtype) +
		# geom_violin(aes(fill=testtype)) +
		geom_boxplot(aes(fill=testtype), outlier.alpha=0, alpha=0.8) +
		geom_jitter(width=0.2, alpha=0.3, size=rel(0.1)) +
		geom_hline(yintercept=0.05, colour="grey65") +
		guides(fill="none") +
		ggtitle("DTU False Discovery Rate (default parameters)") +
		xlab("Replicates per condition") +
		ylab("FDR") +
		scale_y_continuous(expand=c(0,0), breaks=c(0, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06)) +
		coord_cartesian(ylim=c(0,0.065)) +
		gth
)
# Different view of same data.
print( 
	ggplot(fp_default, aes(x=Reps, y=FDR)) +
		facet_grid(. ~ testtype) +
		geom_violin(aes(fill=testtype)) +
		# geom_boxplot(aes(fill=testtype), outlier.alpha=0, alpha=0.8) +
		# geom_jitter(width=0.2, alpha=0.3, size=rel(0.8)) +
		geom_hline(yintercept=0.05, colour="grey65") +
		guides(fill="none") +
		ggtitle("DTU False Discovery Rate (default parameters)") +
		xlab("Replicates per condition") +
		ylab("FDR") +
		scale_y_continuous(expand=c(0,0), breaks=c(0, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06)) +
		coord_cartesian(ylim=c(0,0.065)) +
		gth
)
```

The false positives performance stays well under 0.05, with a median FDR of 0.03 at the 3-replicate level and only a handful of iterations
Higher numbers of replicates per condition quickly drive the FDR further down.


## FDR: Effect size VS replication


```{r}
# Genes.
FDRsg <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		slice <- data.frame("FDR" = mega[[rl]][[b]]["gene", dq, , dp, da] / elig[[rl]][[b]][da, "eligen"],
								"dprop" = dprop_range)
		return(slice)
	}) )
	meanfp <- data.frame( "FDR" = sapply(dprop_range, function(d) { mean(fp[fp$dprop==d, "FDR"]) }),
						  "dprop" = as.numeric(dprop_range),
						  "type" = c("genes"),
						  "Replication" = c(rl)
	)
	return(meanfp)
}, mc.cores = 7) )
# Transcripts.
FDRst <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		slice <- data.frame("FDR" = mega[[rl]][[b]]["transc", dq, , dp, da] / elig[[rl]][[b]][da, "elitra"],
								"dprop" = dprop_range)
		return(slice)
	}) )
	meanfp <- data.frame( "FDR" = sapply(dprop_range, function(d) { mean(fp[fp$dprop==d, "FDR"]) }),
						  "dprop" = as.numeric(dprop_range),
						  "type" = c("transcripts"),
						  "Replication" = c(rl)
	)
	return(meanfp)
}, mc.cores = 7) )

FDR_dq <- rbind(FDRsg, FDRst)

print(
	ggplot(FDR_dq, aes(x=dprop, y=FDR, colour=Replication)) +
		facet_grid(. ~ type) + 
		geom_hline(yintercept=0.05, colour="grey65") +
		geom_line() +
		scale_x_continuous(expand=c(0,0)) +
		scale_y_continuous(expand=c(0,0)) +
		coord_cartesian(xlim=c(0, 0.33)) +
		scale_colour_brewer(type='div', palette='RdBu') +
		guides(colour="legend") +
		ggtitle("DTU Mean FDR by effect size") +
		xlab("Effect size threshold") +
		ylab("FDR") +
		gth
)
```

Larger effect sizes are detected with lower false positives. A higher number of replicates drives down the false positives for all effect sizes,
allowing the effect size threshold to be lowered.

## FDR: The main 3 criteria.

### FDR: effect size VS reproducibility thresholds

Other parameters with default values.

```{r}
FDRsg <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		df <- do.call("rbind", lapply(qrep_range, function(q) {
			slice <- data.frame("FDR" = mega[[rl]][[b]]["gene", q, , dp, da] / elig[[rl]][[b]][da, "eligen"],
								"qrep" = c(q),
								"dprop" = dprop_range,
								"type" = c("genes"),
								"reps" = c(rl))
			return(slice)
		}) )
		return(df)
	}) )
	return(fp)
}, mc.cores = 7) )

FDRst <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		df <- do.call("rbind", lapply(qrep_range, function(q) {
			slice <- data.frame("FDR" = mega[[rl]][[b]]["transc", q, , dp, da] / elig[[rl]][[b]][da, "elitra"],
								"qrep" = c(q),
								"dprop" = dprop_range,
								"type" = c("transcripts"),
								"reps" = c(rl))
			return(slice)
		}) )
		return(df)
	}) )
	return(fp)
}, mc.cores = 7) )

FDR_dq <- rbind(FDRsg, FDRst)

print(
	ggplot(FDR_dq, aes(x=as.numeric(dprop)/100, y=qrep)) +
		facet_grid(type ~ reps) +
		geom_tile(aes(fill=FDR)) +
		scale_fill_gradient(low="white", high="red") +
		guides(fill="colorbar") +
		ggtitle("DTU FDR by replication level") +
		xlab("Effect size threshold") +
		ylab("Quantification reproducibility threshold") +
		theme(strip.background= element_rect(fill= "grey95"),
	        axis.text.x= element_text(angle= 90),
			panel.grid.major= element_line(colour = "grey95"),
	        panel.grid.minor= element_line(colour = "grey95", size=0.5),
	        panel.background= element_rect(fill = "grey98") )
)
```

The influence of minimum effect is more pronounced for the gene-level test, as it is more sensitive. Most of the effect size's influence happens below the value of 0.05.
The quantification reproducibility threshold appears to have little effect until the very top of its range. Although it is highly correlated high quality replicate data that was used, it is not clear why correlation between the samples would influence Salmon's ability to reproducibly quantify the transcripts in any single sample.
Higher replication compresses the rate at which FDR drops with higher effect size thresholds (less "haze"), but the minimum threshold value of 0.05 remains the same.

### FDR: effect size VS p-value

Other parameters with default values.

```{r}
FDRsg <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		df <- do.call("rbind", lapply(p_range, function(p) {
			slice <- data.frame("FDR" = mega[[rl]][[b]]["gene", dq, , p, da] / elig[[rl]][[b]][da, "eligen"],
								"pval" = c(p),
								"dprop" = dprop_range,
								"type" = c("genes"),
								"reps" = c(rl))
			return(slice)
		}) )
		return(df)
	}) )
	return(fp)
}, mc.cores = 7) )

FDRst <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		df <- do.call("rbind", lapply(p_range, function(p) {
			slice <- data.frame("FDR" = mega[[rl]][[b]]["transc", dq, , p, da] / elig[[rl]][[b]][da, "elitra"],
								"pval" = c(p),
								"dprop" = dprop_range,
								"type" = c("transcripts"),
								"reps" = c(rl))
			return(slice)
		}) )
		return(df)
	}) )
	return(fp)
}, mc.cores = 7) )

FDR_dq <- rbind(FDRsg, FDRst)

print(
	ggplot(FDR_dq, aes(x=pval, y=as.numeric(dprop)/100)) +
		facet_grid(type ~ reps) +
		geom_tile(aes(fill=FDR)) +
		scale_fill_gradient(low="white", high="blue") +
		guides(fill="colorbar") +
		ggtitle("DTU FDR by replication level") +
		xlab("Significance cutoff") +
		ylab("Effect size threshold") +
		theme(strip.background= element_rect(fill= "grey95"),
	        axis.text.x= element_text(angle= 90),
			panel.grid.major= element_line(colour = "grey95"),
	        panel.grid.minor= element_line(colour = "grey95", size=0.5),
	        panel.background= element_rect(fill = "grey98") )
)
```

The effect size has a more decisive role than significance in the classification of DTU. Most of its influence
takes place up to the value of 0.05, with little affect above 0.10. One must take into account that these results are
obtained from comparisons between samples of the same condition, so large effect sizes are unlikely to be present to begin with.
The p-value alone, on the other hand, produces considerable FPs even at its strictest cutoff unless combined with higher effect size thresholds.
These observations are not affected by the replication level.

### FDR: reproducibility VS p-value

Other parameters with default values.

```{r}
FDRsg <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		df <- do.call("rbind", lapply(p_range, function(p) {
			slice <- data.frame("FDR" = mega[[rl]][[b]]["gene", , dd, p, da] / elig[[rl]][[b]][da, "eligen"],
								"pval" = c(p),
								"qrep" = qrep_range,
								"type" = c("genes"),
								"reps" = c(rl))
			return(slice)
		}) )
		return(df)
	}) )
	return(fp)
}, mc.cores = 7) )

FDRst <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		df <- do.call("rbind", lapply(p_range, function(p) {
			slice <- data.frame("FDR" = mega[[rl]][[b]]["transc", , dd, p, da] / elig[[rl]][[b]][da, "elitra"],
								"pval" = c(p),
								"qrep" = qrep_range,
								"type" = c("transcripts"),
								"reps" = c(rl))
			return(slice)
		}) )
		return(df)
	}) )
	return(fp)
}, mc.cores = 7) )

FDR_dq <- rbind(FDRsg, FDRst)

print(
	ggplot(FDR_dq, aes(x=pval, y=qrep)) +
		facet_grid(type ~ reps) +
		geom_tile(aes(fill=FDR)) +
		scale_fill_gradient(low="white", high="orange") +
		guides(fill="colorbar") +
		ggtitle("DTU FDR by replication level") +
		xlab("Significance cutoff") +
		ylab("Reproducibility threshold") +
		theme(strip.background= element_rect(fill= "grey95"),
	        axis.text.x= element_text(angle= 90),
			panel.grid.major= element_line(colour = "grey95"),
	        panel.grid.minor= element_line(colour = "grey95", size=0.5),
	        panel.background= element_rect(fill = "grey98") )
)
```

Higher replication lowers the false positives. The reproducibility again shows little effect. The significance cutoff does not
have much dynamic range in terms of FDR. the number of replicates is the clear most important factor here.

## FDR: the noise filters

### FDR: abundance VS effect size

Other parameters with default values.

```{r}
FDRsg <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		df <- do.call("rbind", lapply(abund_range, function(a) {
			slice <- data.frame("FDR" = mega[[rl]][[b]]["gene", dq, , dp, a] / elig[[rl]][[b]][a, "eligen"],
								"abund" = c(a),
								"dprop" = dprop_range,
								"type" = c("genes"),
								"reps" = c(rl))
			return(slice)
		}) )
		return(df)
	}) )
	return(fp)
}, mc.cores = 7) )

FDRst <- do.call("rbind", mclapply(replevels, function(rl) {
	fp <- do.call("rbind", lapply(seq.int(1,bootnum,1), function(b) {
		df <- do.call("rbind", lapply(abund_range, function(a) {
			slice <- data.frame("FDR" = mega[[rl]][[b]]["transc", dq, , dp, a] / elig[[rl]][[b]][a, "elitra"],
								"abund" = c(a),
								"dprop" = dprop_range,
								"type" = c("transcripts"),
								"reps" = c(rl))
			return(slice)
		}) )
		return(df)
	}) )
	return(fp)
}, mc.cores = 7) )

FDR_dq <- rbind(FDRsg, FDRst)

print(
	ggplot(FDR_dq, aes(x=abund, y=as.numeric(dprop)/100)) +
		facet_grid(type ~ reps) +
		geom_tile(aes(fill=FDR)) +
		scale_fill_gradient(low="white", high="purple") +
		guides(fill="colorbar") +
		ggtitle("DTU FDR by replication level") +
		xlab("Abundance threshold") +
		ylab("Effect size threshold") +
		theme(strip.background= element_rect(fill= "grey95"),
	        axis.text.x= element_text(angle= 90),
			panel.grid.major= element_line(colour = "grey95"),
	        panel.grid.minor= element_line(colour = "grey95", size=0.5),
	        panel.background= element_rect(fill = "grey98") )
)
```

Pre-filtering low-abundance transcripts shows a discernible reduction in FDR, with high values allowing lower effect size threshold.
But looking back at the graph of eligible features at the begining of this workbook shows that this benefit comes at the cost of greatly 
reduced number of testable-features. Also, below the critical point of 0.05 for the effect size, more aggressive abundance filtering leads
to increased presence of FPs, particularly for the gene-level test.

