---
title: 'RATs: Plots'
author: "Kimon Froussios"
date: "09 FEB 2018"
output:
  html_document:
    keep_md: yes
    theme: readable
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
vignette: >
  \usepackage[utf8]{inputenc}
  %\VignetteIndexEntry{RATs 4: Plots} 
  %\VignetteEngine{knitr::rmarkdown} 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

***

```{r, include=FALSE}
library(rats)
```

```{r, eval=FALSE}
library(rats)
```


Set up an example.

```{r}
# Simulate some data.
simdat <- sim_boot_data(clean=TRUE) 
# For convenience let's assign the contents of the list to separate variables
mycond_A <- simdat[[2]]   # Simulated bootstrapped data for one condition.
mycond_B <- simdat[[3]]   # Simulated bootstrapped data for other condition.
myannot <- simdat[[1]]    # Transcript and gene IDs for the above data.

# Call DTU
mydtu <- call_DTU(annot= myannot, verbose= FALSE,
                  boot_data_A= mycond_A, boot_data_B= mycond_B,
                  dprop_thresh=0.1, qboot=TRUE, rboot=FALSE)
```


***


# Visualisation of results


The RATs output object provides a host of information and we encourage users to familiarize themselves with it. 
But a good plot is worth a thousand numbers. RATs provides a non-exhaustive set of basic visualisation options for the results.


## Isoform abundances for a given gene

This function allows you to visualise what's going on in any particular gene. Both the absolute counts and the relative 
proportions are plotted for each transcript. This is a very useful function for inspecting a gene of interest. It enables 
quick visual evaluation of the dispersion of the replicate measurements, the magnitude of the proportion change, the 
presence of outliers, and the consistency among the replicates.

The gene plot can encode the transcript-level DTU outcome, the replicate of origin, the condition or the isoform. These are
represented as line colour, box fill or point shape. The default choices may not please everyone, so these are all controlled 
by parameters (see Plot Customisation later in this vignette).

There are several styles for this plot, depending on your preferences. The default plot format includes the most information:

```{r}
# Grouping by condition (DEFAULT):
plot_gene(mydtu, "MIX6", style="bycondition")
```

The top two facets show the absolute abundances (counts) of the isoforms as supplied in the input, the bottom two show the correspondng 
relative abundances (proportions). These are split by condition: The left two facets refer to one condition, the right two ones to the 
other condition. The boxplots describe the abundance measurements of each isoform across replicates. 
The actual measurements are overlayed as points and connected by lines so as to group the set of isoform abundances measured in each replicate. 
This highlights the level of consistency between the replicates. The points and lines are placed slightly off-centre to prevent replicates 
masking one another when the measurements are very similar. 

The colours of the replicates are recycled between the two conditions, to reducethe required palette and boost contrast. 
So in the example, there are 4 samples: controls-1, controls-2, patients-1 and patients-2.

In the past, the fill colour encoded the DTU outcome, but to keep things more consistent with the other styles
the DTU is now encoded as point shape while the fill colour echoes the condition.
In this example, isoforms `.c1` and `.c2` change significantly, isoforms `.c3`, `.c4` and `.nc` do not change significantly 
and isoform `.d` could not be tested.

This structure is great for seeing changes in the isoform distribution profile of a gene. A more minimalist version of this, without the boxpltos,
can also be obtained:

```{r, eval=FALSE}
# Grouping by condition (minimalist):
plot_gene(mydtu, "MIX6", style="lines")
```

When there are many replicates or many isoforms, it may be preferable to group abundances by isoform, making individual comparisons easier:

```{r}
# Grouping by isoform:
plot_gene(mydtu, "MIX6", style="byisoform")
```

### Structure of a given gene

With the use of the third-party `ggbio` package, one can plot gene models from GRanges/GRangesList objects. This can be useful in interpreting transcript expression and DTU, but getting an annotation sliced up in a suitable way can be a little tedious. RATs offers a helper function to create an appropriate collection of GRanges/GRangesList objects from a GTF/GFF3 file. The collection is structured as a list (indexed by gene ID), where each element is a GRangesList object containing a GRange per transcript of that gene.

```{r eval=FALSE}
models <- annot2models('/my/annotation/file.gtf')
library(ggbio)
# This will plot the structure described by the 3rd entry for the given gene.
autoplot(models[['mygeneID']][[3]])
```

For more options and more refined plots, refer to the `ggbio` documentation.

## Overview plots

Our simulated dataset is too small to properly demonstrate what these plots typically look like.
So, instead, each one is accompanied by a static image of the plot created with a real and much larger dataset.

Possibly the most common plot in differential expression is the volcano plot, which plots the effect size against 
the statistical significance. The thresholds are also shown, although the default significance threshold of 0.05
is very low (hint: no lines are drawn for the axes).

```{r, eval=FALSE}
# Proportion change VS transcript-level significance. Each point is a transcript
plot_overview(mydtu, type="tvolcano")

# This can also be plotted for genes, by using the largest isoform effect size as proxy.
plot_overview(mydtu, type="gvolcano")
```

This is what these look like on a larger dataset:
![Transcript significance VS effect size](./figs/tvolcano.jpg)
![Gene significance VS largest effect size among isoforms](./figs/gvolcano.jpg)

You can also get density histograms for the volcanos (the Y axis is square-root compressed):

```{r, eval=FALSE}
# Distribution of proportion change.
plot_overview(mydtu, type="dprop")

# Distribution of largest isoform proportion change per gene.
plot_overview(mydtu, type="maxdprop")
```

This is what these look like on a larger dataset:
![Effect size distribution](./figs/dprop.jpg)
![Largest isoform effect size per gene](./figs/maxdprop.jpg)

The gene version of the effect size distribution is often bi-modal with the positive peak being larger than the negative, 
because when a gene has isoforms with the same absolute effect size (for example genes with two isoforms) 
the positive value is deterministically selected.

Although fold-changes of transcript expression are not tied to DTU, if your input counts are suitably normalised between conditions,
you can plot the traditional FC volcano and the relationship between FC and proportion change.

```{r, eval=FALSE}
# Proportion change VS transcript-level significance. Each point is a transcript
plot_overview(mydtu, type="fcvolcano")

# This can also be plotted for genes, by using the largest isoform effect size as proxy.
plot_overview(mydtu, type="fcVSdprop")
```

![Fold change VS significance](./figs/fcvolcano.jpg)
![Fold change VS proportion change](./figs/fcvsdprop.jpg)


## Diagnostic plots

Currently there is only one plot type in this category.

```{r}
# Matrix of pairwise Pearson's correlations among samples.
plot_diagnostics(mydtu, type='cormat') # Default type.
```

What you want to see here, is a nice separation between the samples of different conditions.
Anomalies in this plot may indicate batch effects or mislabelled samples.


## Interactive plots

If you prefer picking points from a plot rather than sorting through tables, the gene-level volcano plot is also available through 
an interactive `shiny` app, that pulls up the relevant `Gene` and `Transcript` results and the isoform abundance plot 
for any volcano point you select.

```{r, eval=FALSE}
# Start the interactive volcano plot.
plot_shiny_volcano(mydtu)
```

1. Hovering over a point (or cluster of points) will show the gene IDs and some summary info.
2. Clicking on a point will display the all the available information calculated for the gene and its isoforms, as
well as draw the isoform abundance plot for the gene.

When you finish exploring the volcano plot, close the popup window or stop the shiny runtime in order to return to your R terminal.


# Plot customisation

## Customisation of the gene plot

### Change information layers

The `fillby`, `colourby` and `shapeby` parameters can be respectively used to control which information layers
are encoded as fill, line/point colour, and point shape. Possible values are `c("isoform", "condition", "DTU", "none", "replicate")`.
Be aware that some combinations of plot style and information layers are not possible. If safe to do so these will be silently ignored,
otherwise an error message will appear.

```{r}
# For a less busy look, any of the information layers can be disabled.
plot_gene(mydtu, "MIX6", style="byisoform", colourby="none", shapeby="none")
```

The following command will approximate the older version of the gene plot (where DTU determined the fill colour):

```{r}
plot_gene(mydtu, "MIX6", fillby="DTU", shapeby="none")
```

### Change colour code

```{r}
# Colour codes can be customised by specifying new values to the
# corresponding parameters.
plot_gene(mydtu, "MIX6", style="bycondition", fillby="condition", 
          condcolvec=c("magenta", "cyan"),
          replcolvec=c("green", "darkgreen"))
```

* `isofcolvec` - Colour vector for isoform highlighting. Used to build a colorRampPalette.
* `dtucolvec` - Colour vector for DTU highlighting.
* `condcolvec` - Colour vector for condition highlighting.
* `replcolvec` - Colour vector for replicate highlighting. Used to build a colorRampPalette.
* `nonecol` - Colour to use when no colour coding is wanted.


## General customisations for all plots

You can save any of the gene and overview plots as a `ggplot2` object and use ggplot2 manipulations on it, such as
the axis tick marks, axis values, labels, titles, colours... Consult the [ggplot2](http://ggplot2.org)
documentation for more help on these.


***


# Contact information

The `rats` R package was developed within [The Barton Group](http://www.compbio.dundee.ac.uk) at [The University of Dundee](http://www.dundee.ac.uk)
by Dr. Kimon Froussios, Dr. Kira Mour??o and Dr. Nick Schurch.

To **report problems** or **ask for assistance**, please raise a new issue [on the project's support forum](https://github.com/bartongroup/Rats/issues).
Providing a *reproducible working example* that demonstrates your issue is strongly encouraged to help us understand the problem. Also, be sure 
to **read the vignette(s)**, and browse/search the support forum before posting a new issue, in case your question is already answered there.

Enjoy!

![](./figs/rats_logo.png)


